### PR TITLE
Settings nvs backend, settings/src test integration

### PR DESCRIPTION
**Pull request by kapi-no**
Blocked by: 
~~#18891~~
~~#18924~~

### CODE REVIEW

**Comment by Laczen** [comment]
@kapi-no, I am a bit puzzled by seeing base64 encoding. Is this supported by the nvs backend ? If not, you could just have a raw test for nvs.

**Comment by kapi-no** [comment]
@Laczen, I think that the base64 encoded data are prepared before they are passed to the NVS backend. The backend just receives a raw buffer, so you don't need any more logic in the backend to support base64 encoding.

I am aware that base64 is going to be deprecated in the future, but I added the tests for it nonetheless because they required no effort from my side. I reused common tests that are run for other backends. They should be removed when we get rid of base64 support in the Settings.

**Comment by Laczen** [comment]
@kapi-no, could you recheck this, I think the base64 encoding is only supported in nffs and fcb using the settings_line() method that defines how data is saved (including using base64 or not). The nvs backend does not use the settings_line method but directly saves the data in nvs using a different method (to avoid double writing of variable names), it does not have the need or the possibility to store using base64 encoding. I agree that you should reuse the common tests that are used for the other backends, but for the nvs backend it should only do the raw version testing. It would be confusing for users to do a test with base64 encoding while it really does nothing.

**Comment by kapi-no** [comment]
@Laczen, okay, I will remove it.

Although I didn't analyze base64 related tests, it seems quite strange that they pass for the NVS backend.

**Review by Laczen — APPROVED** [review]

**Review by rakons — APPROVED** [review]

**Comment by rakons** [comment]
It passes base64 tests because enabling or disabling base64 is only a settings flag. The test is not checking the internal format of the data.

**Review by MaureenHelm — DISMISSED** [review]

File: tests/subsys/settings/nvs/raw/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+tests:
+  system.settings.nvs:
+    platform_whitelist: qemu_x86
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: sanitycheck is also used for local hardware testing using the `--device-testing` flag, and whitelists like this will prevent the test from being executed on hardware. We really need to avoid whitelists whenever possible

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: okay, I see. If the test for qemu_86 is still picked up by the CI without this whitelist entry, I can get rid of it

**Review by nvlsianpu — APPROVED** [review]

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: I removed the ```platform_whitelist``` attribute

**Comment by kapi-no** [comment]
I set up an artificial failure in one of the tests just to see if this is actually executed in the Zephyr CI.

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: After I removed ```platform_whitelist```, the ```ztest.h``` file is not visible during compilation in the Zephyr CI tests. Is it related or do I need to rebase?

**Comment by nvlsianpu** [comment]
@kapi-no ^^ yo u can remove this failure.

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Rebase didn't help.

I brought back this attribute since the Zephyr CI failed without having it defined in the yaml file.

**Comment by kapi-no** [comment]
@nvlsianpu, done. This PR is ready for merging

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Please figure out why `ztest.h` isn't being found. `qemu_x86` isn't the only platform that runs tests

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Apparently, the Zephyr CI scripts require this attribute to prepare the environment for test suite execution. Could you give me an example of the test without ```platform_whitelist```?

[the remark this answers is not in the document]
Comment (MaureenHelm) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: `tests/drivers/spi/spi_loopback/testcase.yaml`

**Comment by nvlsianpu** [comment]
@carlescufi Can this be merged?

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: AFIK without specifying qemu we can't build test as a ztes header is missing.

File: tests/subsys/settings/nvs/raw/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(NONE)
```
Comment (nashif) [inline tests/subsys/settings/nvs/raw/CMakeLists.txt]: can you please use something else other than NONE.

File: tests/subsys/settings/nvs/raw/prj_qemu_x86.conf [hunk 3]
```diff
@@ -0,0 +1,11 @@
+CONFIG_ZTEST=y
```
Comment (nashif) [inline tests/subsys/settings/nvs/raw/prj_qemu_x86.conf]: how is this difference from prj_qemu.conf?

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nashif) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: there is something wrong here. ztest has nothing to do with qemu_x86

**Review by nashif — DISMISSED** [review]

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlescufi) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Yes, something is off. The test that @MaureenHelm referred to does not have a whitelist and also includes `ztest.h`:
https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/drivers/spi/spi_loopback/src/spi.c#L15
@kapi-no can you check if there's another difference between yours and the spi one?

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Well, the spi test project specifies that it requires a platform with SPI capability:

https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/drivers/spi/spi_loopback/testcase.yaml#L3

Probably, the SPI test is not generated & run on ```qemu_x86```, because it has no such capability.

To be clear, if I remove the ```platform_whitelist``` attribute, I am still able to generate a valid project for ```qemu_x86``` with cmake. The issue appears when the project is generated in the Zephyr CI environment that is prepared by ```scripts/ci/run_ci.sh```. 

I will try to look into this when I have some spare time, but I suspect that it is some kind of general issue for all tests that run on ```qemu_x86```.

File: tests/subsys/settings/nvs/raw/CMakeLists.txt (hunk 2 above)
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/CMakeLists.txt]: sure, I will correct it in the next patch

File: tests/subsys/settings/nvs/raw/prj_qemu_x86.conf (hunk 3 above)
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/prj_qemu_x86.conf]: there is no difference, I will remove it in the next patch

File: tests/subsys/settings/nvs/raw/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: I investigated this a bit more and the tests fail actually on all platforms except ```qemu_x86``` in the current state. This is because there is only one ```prj_qemu_x86.conf``` file that is used for ```qemu_x86```. Other platforms use default configuration which does not include Ztest framework.

However, when I defined a default configuration for all platforms using ```prj.conf```, the Zephyr CI failed during the previous rebase:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/51146/4/console
Various platforms lacked necessary DT definitions (e.g. ```DT_FLASH_AREA_STORAGE_ID``` / ```DT_FLASH_ERASE_BLOCK_SIZE``` for ```qemu_riscv32```, ```qemu_riscv64``` or ```qemu_x86_64```)

I question the need to have the settings NVS backend tests run on all platforms (192 more tests). These tests are supposed to verify the Settings module logic - not the underlying flash drivers.

To sum up, I think it's better to use ```platform_whitelist``` attribute and have single test executed for ```qemu_x86```. Do you have any objections?

[the remark this answers is not in the document]
Comment (MaureenHelm) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Yes, I still object to the whitelist. You can add `depends_on: nvs` to filter out boards that don't have the necessary DT definitions. See `samples/subsys/nvs/sample.yaml`

```
$ git grep nvs -- boards/*/*/*.yaml
boards/arm/bl652_dvk/bl652_dvk.yaml:  - nvs
boards/arm/disco_l475_iot1/disco_l475_iot1.yaml:  - nvs
boards/arm/efm32hg_slstk3400a/efm32hg_slstk3400a.yaml:  - nvs
boards/arm/efm32pg_stk3402a/efm32pg_stk3402a.yaml:  - nvs
boards/arm/efm32wg_stk3800/efm32wg_stk3800.yaml:  - nvs
boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.yaml:  - nvs
boards/arm/efr32mg_sltb004a/efr32mg_sltb004a.yaml:  - nvs
boards/arm/frdm_k64f/frdm_k64f.yaml:  - nvs
boards/arm/frdm_k82f/frdm_k82f.yaml:  - nvs
boards/arm/nrf51_pca10028/nrf51_pca10028.yaml:  - nvs
boards/arm/nrf51_pca10031/nrf51_pca10031.yaml:  - nvs
boards/arm/nrf52_pca10040/nrf52_pca10040.yaml:  - nvs
boards/arm/nucleo_f091rc/nucleo_f091rc.yaml:  - nvs
boards/arm/nucleo_f334r8/nucleo_f334r8.yaml:  - nvs
boards/arm/nucleo_l432kc/nucleo_l432kc.yaml:  - nvs
boards/arm/nucleo_l4r5zi/nucleo_l4r5zi.yaml:  - nvs
boards/arm/sensortile_box/sensortile_box.yaml:  - nvs
boards/arm/stm32f3_disco/stm32f3_disco.yaml:  - nvs
```

[the remark this answers is not in the document]
Comment (nvlsianpu) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @kapi-no  I'm in line with you - this test is for testing whether software correctly uses other software and SUT logic.

I vote for keeping the white-list here.

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: I added NVS capability to the ```qemu_x86``` board so that we can run the test included in this PR on the  ```qemu_x86``` board using:
```depends on: nvs```
syntax instead of:
```platform_whitelist ```

I was forced to add separate ```prj_qemu_x86.conf``` to the ```samples/subsys/nvs/``` sample due to the new NVS capability of ```qemu_x86``` board.

[the remark this answers is not in the document]
Comment (Laczen) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @kapi-no, nice work.

I think you should remove prj_qemu_x86.conf from the nvs sample, the sample uses reboot, which is not supported by qemu_x86.

[the remark this answers is not in the document]
Comment (MaureenHelm) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: Thanks for removing the whitelist. You probably want to add a min_ram property to the testcase.yaml to filter out platforms that don't have enough ram. This will help shippable to pass. For example:
https://github.com/zephyrproject-rtos/zephyr/blob/e833fafd4ba3e7967b32a0c7226d692b48eb5728/tests/net/all/testcase.yaml#L4-L7

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: NVS tests for following boards have failed:
- ```out-3nd-pass/efm32hg_slstk3400a/tests/subsys/settings/nvs/raw/system.settings.nvs```
- ```out-3nd-pass/nucleo_f334r8/tests/subsys/settings/nvs/raw/system.settings.nvs```
because of SRAM overflow
- ```out-3nd-pass/qemu_x86/samples/subsys/nvs/sample.subsys.nvs```
because of timeout. This is sample and I think that framework expects log output in the test format.

```qemu_x86``` is a special type of board, because it is not only compiled but also tested in the runtime. I guess it cannot declare capabilities that are used in samples, because we will get CI failures.

I think we should stick to ```platform_whitelist``` in this case. @MaureenHelm, would you agree to that?

The topic of this PR is not to change how the tests for settings are defined. Apart from that, it is a bit inconsistent to have one settings test using ```depends on``` syntax and other ones using ```platform whitelist``` (fcb, nffs, functional, etc.).

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @Laczen, I cannot do it since the sample would be built in the default configuration (prj.conf) for ```qemu_x86``` and then the Zephyr CI will fail during compilation. However, I agree with you that we should get rid of ```qemu_x86``` target for the NVS sample.

@MaureenHelm, thanks for the tips on how to solve SRAM overflow issue, but I will still have one failing test:
```out-3nd-pass/qemu_x86/samples/subsys/nvs/sample.subsys.nvs```
This is sample that is now built for  ```qemu_x86``` , because I declared ```nvs``` capability for this board. ```qemu_x86``` is a special type of board, because it is not only compiled but also tested in the runtime. That's why I think that framework expects log output in the test format. Is it possible to somehow blacklist ```qemu_x86``` in the NVS sample?

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @MaureenHelm, please provide a suggestion on how to solve this last issue so that you can approve this PR. If there is no easy way to solve this, I am planning to use ```platform_whitelist``` attribute.

[the remark this answers is not in the document]
Comment (Laczen) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @kapi-no, is it not possible to add depends on: reboot in the nvs sample ?

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @Laczen, I think there is no such capability

[the remark this answers is not in the document]
Comment (MaureenHelm) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: You can use `platform_exclude` to blacklist `qemu_x86` in the nvs sample

[the remark this answers is not in the document]
Comment (kapi-no) [reply tests/subsys/settings/nvs/raw/testcase.yaml]: @MaureenHelm, thanks. This is resolved now

**Comment by kapi-no** [comment]
@nvlsianpu, @carlescufi, I believe this PR is finally ready for merging

**Comment by carlescufi** [comment]
@MaureenHelm @nashif are you happy with this in its current state?

**Review by MaureenHelm — APPROVED** [review]
